roachtest: add validation to schema change random test

spaskob opened the pull request:
We add at new random operation to the suite that is a no-op if all
descriptors validate and fails with details about invalid objects
otherwise.

Release note: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55162)
<!-- Reviewable:end -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55162)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @spaskob)

___
*[pkg/workload/schemachange/schemachange.go, line 172 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55162#-MIypcB96lOE9aJGo3f4:-MIypcB96lOE9aJGo3f5:b-pnmy3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a50d34256f49e772ecfecb0f7830f15ae580d4d2/pkg/workload/schemachange/schemachange.go#L172)):*
> ```Go
> 	setColumnType:     1,
> 	insertRow:         1,
> 	validate:          10, // validate more often
> ```

nit: this feels pretty high. I don't anticipate that corruptions which occur are going to get automatically rectified. Given that, I feel like it might be okay to not spend so much time doing this. What about 2?
___
*[pkg/workload/schemachange/schemachange.go, line 877 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55162#-MIyqTb5BMpDEBj5ncQs:-MIyqTb5BMpDEBj5ncQt:b4s6atn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a50d34256f49e772ecfecb0f7830f15ae580d4d2/pkg/workload/schemachange/schemachange.go#L877)):*
> ```
> 
> 		return validateStmt, nil
> 	}
> ```
I think this is going to lead to a missed row. I think the right way to do this will be to check to see if `len(errs) == 0` .
___
*[pkg/workload/schemachange/schemachange.go, line 879 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55162#-MIyqg4JD2NET8xl-RG8:-MIyqg4JD2NET8xl-RG9:b52g80i) ([raw file](https://github.com/cockroachdb/cockroach/blob/a50d34256f49e772ecfecb0f7830f15ae580d4d2/pkg/workload/schemachange/schemachange.go#L879)):*
> ```Go
> 	}
> 
> 	errs := make([]string, 0)
> ```

nit: the idiom for this would just be `var errs []string`
___
*[pkg/workload/schemachange/schemachange.go, line 890 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55162#-MIyqixtBg8NoM4o_nln:-MIyqixtBg8NoM4o_nlo:b-awcgg3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a50d34256f49e772ecfecb0f7830f15ae580d4d2/pkg/workload/schemachange/schemachange.go#L890)):*
> ```Go
> 			fmt.Sprintf("id %d, db %s, schema %s, name %s: %s", id, dbName, schemaName, objName, errStr),
> 		)
> 	}
> ```

can you check the `rows.Err()` after the loop. 


<!-- Sent from Reviewable.io -->

spaskob reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55162)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/workload/schemachange/schemachange.go, line 877 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55162#-MIyqTb5BMpDEBj5ncQs:-MJ0ckZG1urMKOKmSTvV:b-jjkv20) ([raw file](https://github.com/cockroachdb/cockroach/blob/a50d34256f49e772ecfecb0f7830f15ae580d4d2/pkg/workload/schemachange/schemachange.go#L877)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 
> 		return validateStmt, nil
> 	}
> ```
I think this is going to lead to a missed row. I think the right way to do this will be to check to see if `len(errs) == 0` .


</blockquote></details>

I see...`Next` advances the iterator, I thought that `Scan` would do that - thanks for the catch.
___
*[pkg/workload/schemachange/schemachange.go, line 890 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55162#-MIyqixtBg8NoM4o_nln:-MJ0e8QXE6ZSoXLW3k-9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a50d34256f49e772ecfecb0f7830f15ae580d4d2/pkg/workload/schemachange/schemachange.go#L890)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

can you check the `rows.Err()` after the loop. 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Remove the `snappy` change and <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55162#-:-MJ79Xg-1kjBKf1mRuTH:bkciy20)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @spaskob)

___
*[c-deps/snappy, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55162#-MJ79P0HCbvGV3Lii6-a:-MJ79P0IAfPy2Oo20kXZ:b-suoyd7) ([raw file](https://github.com/cockroachdb/cockroach/blob/26bcbbcca74d5a68e8d4c77a479c538d395ab1a6/c-deps/snappy#L1)):*
> ```
> Subproject commit b02bfa754ebf27921d8da3bd2517eab445b84ff9
> ```

I suspect this is unintentional


<!-- Sent from Reviewable.io -->

Commit:
roachtest: add validation to schema change random test

We add at new random operation to the suite that is a no-op if all
descriptors validate and fails with details about invalid objects
otherwise.

Release note: none.

spaskob commented:
bors r+

spaskob reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55162)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner)

___
*[c-deps/snappy, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55162#-MJ79P0HCbvGV3Lii6-a:-MJ7aam7BbI3Lnbhq7c3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/26bcbbcca74d5a68e8d4c77a479c538d395ab1a6/c-deps/snappy#L1)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I suspect this is unintentional
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

spaskob commented:
bors r+
